File: server/controllers/atsController.js
const { ProfileRepository, HistoryRepository } = require('../models/dbStore');
const llmGateway = require('../services/llmGateway');

// @desc    Analyze resume for ATS match against JD
// @route   POST /api/resume/ats
const analyzeATS = async (req, res, next) => {
  try {
    let { resumeText, jobDescription, company, role } = req.body;

    const profile = await ProfileRepository.findOne({ userId: req.user._id });
    if (profile) {
      if (!resumeText) resumeText = profile.resumeText;
      if (!jobDescription) jobDescription = profile.jobDescription;
      if (!company) company = profile.company;
      if (!role) role = profile.role;
    }

    if (!resumeText || resumeText.trim().length === 0) {
      return res.status(400).json({
        success: false,
        message: 'Resume content is required. Please upload your resume first.',
      });
    }

    const prompt = `
Perform a thorough Applicant Tracking System (ATS) scan and match analysis comparing the candidate's Resume against the Job Description (for ${role || 'Target Role'} at ${company || 'Target Company'}).

Provide realistic, accurate, and actionable ATS feedback in valid JSON with this exact schema:
{
  "atsScore": <number between 0 and 100>,
  "keywordMatchPercentage": <number between 0 and 100>,
  "matchedSkills": ["<skill1>", "<skill2>", ...],
  "missingSkills": ["<missing1>", "<missing2>", ...],
  "strengths": ["<strength1>", "<strength2>", "<strength3>"],
  "suggestions": ["<suggestion1>", "<suggestion2>", "<suggestion3>"],
  "summary": "<2-3 sentence executive evaluation of the resume ATS fitness>"
}
`;

    const context = {
      resumeText,
      jobDescription: jobDescription || 'Standard requirements for modern software and tech roles.',
      company,
      role,
    };

    const aiResult = await llmGateway.generateAIResponse({
      prompt,
      context,
      jsonMode: true,
      moduleType: 'ats',
      systemPrompt: 'You are an advanced enterprise Applicant Tracking System (ATS) and Senior Technical Recruiter analyzer.',
    });

    const atsData = aiResult.parsedJson || {
      atsScore: 78,
      keywordMatchPercentage: 72,
      matchedSkills: ['JavaScript', 'React', 'REST APIs', 'Git'],
      missingSkills: ['Docker', 'AWS', 'CI/CD'],
      strengths: ['Clear project breakdown', 'Relevant fullstack skills'],
      suggestions: ['Add quantitative impact metrics', 'Include cloud deployment keywords'],
      summary: 'Solid profile with strong core fundamentals.'
    };

    // Save to user history
    await HistoryRepository.create({
      userId: req.user._id,
      module: 'ATS Scanner',
      company: company || 'General',
      role: role || 'General Tech Role',
      summary: `ATS Score: ${atsData.atsScore}/100 | Keyword Match: ${atsData.keywordMatchPercentage}%`,
      metadata: { atsScore: atsData.atsScore, matchedCount: atsData.matchedSkills?.length || 0 },
    });

    res.status(200).json({
      success: true,
      data: atsData,
      provider: aiResult.provider,
    });
  } catch (error) {
    next(error);
  }
};

module.exports = {
  analyzeATS,
};

File: server/src/main/java/com/nexoffer/controller/AtsController.java
package com.nexoffer.controller;

import com.nexoffer.model.History;
import com.nexoffer.model.Profile;
import com.nexoffer.model.User;
import com.nexoffer.repository.HistoryRepository;
import com.nexoffer.repository.ProfileRepository;
import com.nexoffer.service.LlmGatewayService;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/resume")
public class AtsController {

    private final ProfileRepository profileRepository;
    private final HistoryRepository historyRepository;
    private final LlmGatewayService llmGatewayService;

    public AtsController(ProfileRepository profileRepository,
                         HistoryRepository historyRepository,
                         LlmGatewayService llmGatewayService) {
        this.profileRepository = profileRepository;
        this.historyRepository = historyRepository;
        this.llmGatewayService = llmGatewayService;
    }

    @PostMapping("/ats")
    public ResponseEntity<?> analyzeAts(@AuthenticationPrincipal User user, @RequestBody(required = false) Map<String, String> body) {
        String resumeText = body != null ? body.get("resumeText") : null;
        String jobDescription = body != null ? body.get("jobDescription") : null;
        String company = body != null ? body.get("company") : null;
        String role = body != null ? body.get("role") : null;

        Optional<Profile> profileOpt = profileRepository.findByUserId(user.getId());
        if (profileOpt.isPresent()) {
            Profile p = profileOpt.get();
            if (resumeText == null || resumeText.isBlank()) resumeText = p.getResumeText();
            if (jobDescription == null || jobDescription.isBlank()) jobDescription = p.getJobDescription();
            if (company == null || company.isBlank()) company = p.getCompany();
            if (role == null || role.isBlank()) role = p.getRole();
        }

        if (resumeText == null || resumeText.isBlank()) {
            return ResponseEntity.badRequest().body(Map.of("success", false, "message", "Resume content is required. Please upload your resume first."));
        }

        String prompt = "Perform a thorough Applicant Tracking System (ATS) scan and match analysis comparing the candidate's Resume against the Job Description (for " +
                (role != null ? role : "Target Role") + " at " + (company != null ? company : "Target Company") + ").\n\n" +
                "Provide realistic, accurate, and actionable ATS feedback in valid JSON with this exact schema:\n" +
                "{\n" +
                "  \"atsScore\": <number 0-100>,\n" +
                "  \"keywordMatchPercentage\": <number 0-100>,\n" +
                "  \"matchedSkills\": [\"skill1\", \"skill2\"],\n" +
                "  \"missingSkills\": [\"missing1\", \"missing2\"],\n" +
                "  \"strengths\": [\"strength1\", \"strength2\"],\n" +
                "  \"suggestions\": [\"suggestion1\", \"suggestion2\"],\n" +
                "  \"summary\": \"evaluation summary\"\n" +
                "}";

        Map<String, String> context = new HashMap<>();
        context.put("resumeText", resumeText);
        context.put("jobDescription", jobDescription != null ? jobDescription : "Standard tech requirements");
        context.put("company", company != null ? company : "Target Company");
        context.put("role", role != null ? role : "Software Engineer");

        LlmGatewayService.AiResponse aiResponse = llmGatewayService.generateAIResponse(
                prompt,
                "You are an advanced enterprise Applicant Tracking System (ATS) and Senior Technical Recruiter analyzer.",
                context,
                true,
                "ats"
        );

        Object data = aiResponse.getParsedJson() != null ? aiResponse.getParsedJson() : llmGatewayService.generateFallbackResponse("ats", context);

        // Record history
        historyRepository.save(new History(
                user.getId(),
                "ATS Scanner",
                company != null ? company : "General",
                role != null ? role : "Software Role",
                "ATS Compatibility Scan Completed"
        ));

        return ResponseEntity.ok(Map.of(
                "success", true,
                "data", data,
                "provider", aiResponse.getProvider()
        ));
    }
}
